Add logs agent integ test

varunch77 opened the pull request:
# Description of the issue
We are currently missing tests that ensure that the logs agent is properly sending an entity when sending application logs to CloudWatch Logs.

# Description of changes
`TestWriteLogsWithEntityInfo()` writes logs to CloudWatch and validates that the logs are associated with entities and that all values are translated properly. The three test cases are `IAMRole`, `ServiceInConfig` (which uses the added `config_log_service_name.json` configuration), and `EC2Tags`. The changes also include a suite of helper functions like `ValidateEntity()` to validate the logs and check that the expected entity information is present after querying the published logs. 

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
I ran all tests in the `publish_logs_test.go` file on a newly created EC2 instance with the most recent build of the cloudwatch agent and the end-to-end testing IAM role attached. I also verified that `TestWriteLogsWithEntityInfo` passed for every running order combination of the IAM, EC2, and Service test cases.

Test run: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/12283029637

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -119,9 +161,9 @@ func TestWriteLogsToCloudWatch(t *testing.T) {
 			// ensure that there is enough time from the "start" time and the first log line,
 			// so we don't miss it in the GetLogEvents call
-			time.Sleep(sleepForFlush)
+			time.Sleep(sleepForExtendedFlush)
```
Why not just increase `sleepForFlush`? Do we use `sleepForFlush` anywhere else in the code?

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -139,6 +181,120 @@ func TestWriteLogsToCloudWatch(t *testing.T) {
+
+			ValidateEntity(t, instanceId, instanceId, &end, testCase.expectedEntity)
+
+			f.Close()
```
feel like we can `f.Close()` after we are done writing on line 285

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -332,3 +488,152 @@ func checkData(t *testing.T, start time.Time, lineCount int) {
+	log.Printf("Validating entity for log group: %s, stream: %s", logGroup, logStream)
+
+	logGroupInfo, err := getLogGroup()
+	for _, lg := range logGroupInfo {
```
Is there no way to get a specific log group when you provide the name? 

Seems expensive to get all log groups in an account. We have a ton in our test account..

varunch77 replied:
I don't think there's a way to just search for one log group (why not is beyond me..) but I updated this so it passes in the log group as a prefix to search by (commit [006a057](https://github.com/aws/amazon-cloudwatch-agent-test/pull/433/commits/006a057bf450c4f0f243278f50bde941a01dd2d9))

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -332,3 +488,152 @@ func checkData(t *testing.T, start time.Time, lineCount int) {
+		}
+	}
+	assert.NoError(t, err)
+	begin := end.Add(-sleepForExtendedFlush * 4)
```
Instead of trying to calculate the begin time, can we just pass in the begin time?

varunch77 replied:
Updated this for `ValidateEntity()` (commit [7950725](https://github.com/aws/amazon-cloudwatch-agent-test/pull/433/commits/795072522768ef075f3068f93a5c84642193acc3))

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -332,3 +488,152 @@ func checkData(t *testing.T, start time.Time, lineCount int) {
+func getLogQueryId(logGroup string, since, until *time.Time) (*string, error) {
+	var queryId *string
+	params := &cloudwatchlogs.StartQueryInput{
+		QueryString:  aws.String("fields @message, @entity.KeyAttributes.Type, @entity.KeyAttributes.Name, @entity.KeyAttributes.Environment, @entity.Attributes.PlatformType, @entity.Attributes.EC2.InstanceId"),
```
nit: make this a const

varunch77 replied:
addressed in commit [84cc3d3](https://github.com/aws/amazon-cloudwatch-agent-test/pull/433/commits/84cc3d39def9c67c5a0460105bd990fe712f031e#diff-90e695612350cf075a13140798e4d267b81f04e5d9258b705f170ab76499d804R53)

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -332,3 +488,152 @@ func checkData(t *testing.T, start time.Time, lineCount int) {
+	if !assert.NotZero(t, len(result)) {
+		return
+	}
+	requiredEntityFields := map[string]bool{
```
Not sure I understand why we need this...when we pass in the query:

```
fields @message, @entity.KeyAttributes.Type, @entity.KeyAttributes.Name, @entity.KeyAttributes.Environment, @entity.Attributes.PlatformType, @entity.Attributes.EC2.InstanceId
```

Doesn't the fact that there are results mean that the logs all have the required fields?

varunch77 commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -86,8 +107,29 @@ type cloudWatchLogGroupClassTestInput struct {
+	)
+	if err != nil {
+		fmt.Println("There was an error trying to load default config: ", err)
+		return
```
It didn't, but I changed the behavior so that it will fail now if the default config fails in commit [84cc3d3](https://github.com/aws/amazon-cloudwatch-agent-test/pull/433/commits/84cc3d39def9c67c5a0460105bd990fe712f031e)

varunch77 commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -29,9 +37,19 @@ const (
 	configPathAutoRemoval         = "resources/config_auto_removal.json"
 	standardLogGroupClass         = "STANDARD"
 	infrequentAccessLogGroupClass = "INFREQUENT_ACCESS"
+	cwlPerfEndpoint               = "https://logs.us-west-2.amazonaws.com"
```
I think we may already be constrained by the region being hardcoded elsewhere: https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/util/awsservice/constant.go#L58C1-L60C78. We would need to make edits there but I'm a little wary of what the blast radius of that would look like for our testing

zhihonl replied:
We don't need to hardcode the endpoint here. The code I wrote originally had to use a custom endpoint because it was a beta endpoint but this is the public us-west-2 endpoint which the client will automatically resolve

varunch77 commented:
New run: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/11978587627

zhihonl commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -29,9 +37,20 @@ const (
 	standardLogGroupClass         = "STANDARD"
 	infrequentAccessLogGroupClass = "INFREQUENT_ACCESS"
+	cwlPerfEndpoint               = "https://logs.us-west-2.amazonaws.com"
+	pdxRegionalCode               = "us-west-2"
```
region hardcoding is not needed since entity feature is GA

zhihonl commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -86,8 +108,28 @@ type cloudWatchLogGroupClassTestInput struct {
+		log.Fatalf("Failed to load default config: %v", err)
+	}
+
+	cwlClient = cloudwatchlogs.NewFromConfig(awsCfg, func(o *cloudwatchlogs.Options) {
```
we don't need to create a new client. we have globally available client that already uses us-west-2 as the region" https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/util/awsservice/constant.go#L74

zhihonl commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -86,8 +108,28 @@ type cloudWatchLogGroupClassTestInput struct {
+	cwlClient = cloudwatchlogs.NewFromConfig(awsCfg, func(o *cloudwatchlogs.Options) {
+		o.BaseEndpoint = aws.String(cwlPerfEndpoint)
+	})
+	ec2Client = ec2.NewFromConfig(awsCfg)
```
same as above

zhihonl commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -332,3 +497,154 @@ func checkData(t *testing.T, start time.Time, lineCount int) {
+func ValidateEntity(t *testing.T, logGroup, logStream string, begin, end *time.Time, expectedEntity expectedEntity) {
+	log.Printf("Validating entity for log group: %s, stream: %s", logGroup, logStream)
+
+	logGroupInfo, err := getLogGroup(logGroup)
```
You can just use `IsLogGroupExists` from our common package: https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/util/awsservice/cloudwatchlogs.go#L137

The helper functions that I wrote was because we had to create a new client in us-east-1 for beta purpose but they are no longer needed

zhihonl commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -332,3 +497,154 @@ func checkData(t *testing.T, start time.Time, lineCount int) {
+	queryId, err := getLogQueryId(logGroup, begin, end)
+	assert.NoError(t, err)
+	log.Printf("queryId is " + *queryId)
+	result, err := getQueryResult(queryId)
```
`getLogQueryId` and `getQueryResult` are not needed anymore since we can just use default client now. There's a helper function here: https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/util/awsservice/cloudwatchlogs.go#L202

zhihonl commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -332,3 +497,154 @@ func checkData(t *testing.T, start time.Time, lineCount int) {
 	assert.NoError(t, err)
 }
+
+func ValidateEntity(t *testing.T, logGroup, logStream string, begin, end *time.Time, expectedEntity expectedEntity) {
```
nonblocking but recommended: This function has a lot of overlaps with this function: https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/test/entity/entity_test.go#L241

Maybe good idea to move `ValidateEntity` to a common package and just use for any entity validation tests in the future.

varunch77 commented:
Addressed comments above, including merging `ValidateEntity` with `ValidateLogEntity` and moving the resulting function to the `util` package so it can be used in multiple places for future entity validation tests.

New run: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/12172724382
